src/main/java/org/truffleruby/language/methods/LookupMethodNode.java

```diff
@@ -68,51 +72,83 @@ public InternalMethod lookup(VirtualFrame frame, Object self, String name) {
+            @Cached("createBinaryProfile()") ConditionProfile foundProfile,
+            @Cached("createBinaryProfile()") ConditionProfile publicProfile) {
+
+        // Find the caller class
+        final DynamicObject callerClass;
+        if (ignoreVisibility || onlyLookupPublic) {
+            callerClass = null; // No need to check visibility
+        } else {
+            final InternalMethod method = RubyArguments.tryGetMethod(frame);
 
-    protected DynamicObject metaClass(Object object) {
-        return metaClassNode.executeMetaClass(object);
-    }
+            if (noCallerMethodProfile.profile(method == null)) {
+                callerClass = coreLibrary().getObjectClass();
+            } else if (!isSendProfile.profile(coreLibrary().isSend(method))) {
+                callerClass = callerMetaClassNode.executeMetaClass(RubyArguments.getSelf(frame));
+            } else {
+                FrameInstance instance = getContext().getCallStack().getCallerFrameIgnoringSend();
+                Frame callerFrame = instance.getFrame(FrameInstance.FrameAccess.READ_ONLY);
+                callerClass = callerMetaClassNode.executeMetaClass(RubyArguments.getSelf(callerFrame));
+            }
+        }
 
-    protected MethodLookupResult doLookup(VirtualFrame frame, Object self, String name) {
-        return lookupMethodWithVisibility(getContext(), frame, self, name, ignoreVisibility, onlyLookupPublic);
-    }
+        // Actual lookup
+        final DynamicObject metaClass = metaClass(self);
 
-    @TruffleBoundary
-    protected static MethodLookupResult doLookup(RubyContext context,
-            DynamicObject callerClass, Object receiver, String name,
-            boolean ignoreVisibility, boolean onlyLookupPublic) {
-        if (RubyGuards.isForeignObject(receiver)) {
+        if (metaClass == coreLibrary().getTruffleInteropForeignClass()) {
+            foreignProfile.enter();
             throw new UnsupportedOperationException("method lookup not supported on foreign objects");
         }
 
-        final MethodLookupResult method = ModuleOperations.lookupMethod(context.getCoreLibrary().getMetaClass(receiver), name);
+        final InternalMethod method = ModuleOperations.lookupMethodUncached(metaClass, name);
 
-        if (!method.isDefined()) {
-            return method.withNoMethod();
+        if (foundProfile.profile(method == null || method.isUndefined())) {
+            return null;
         }
 
         // Check visibility
+        if (publicProfile.profile(method.getVisibility() == Visibility.PUBLIC)) {
+            return method;
+        }
+
         if (!ignoreVisibility) {
             if (onlyLookupPublic) {
-                if (!method.getMethod().getVisibility().isPublic()) {
-                    return method.withNoMethod();
-                }
-            } else if (!method.getMethod().isVisibleTo(callerClass)) {
-                return method.withNoMethod();
+                return null;
```

Comment:
It looks like you're profiling every other branch. Is there any reason to skip these?

Reply:
`ignoreVisibility` and `onlyLookupPublic` are already final fields.
I should profile the isVisibleTo() though.